File: backend/services/mandate.py
"""The mandate: the constraints the book is run under, as ONE object.

Every number here was already being enforced. None of them was written down in a
place a reader could find, and the two places parts of them WERE written down
disagreed with each other:

  * `frontend/lib/risk/riskBoard.ts` declared `gross_exposure_pct: 2.0` — 200%
    gross — while `optimizer.py` has enforced `max_gross = 1.0` since ADR-0037.
    The risk board was permitting leverage the sizer structurally cannot produce,
    which is a published risk limit disagreeing with the published book (ADR-0123).
  * That same board looked up `max_single_name_weight` / `max_sector_weight` /
    `max_geo_weight` in `scoring_config`, where they had never been seeded, so all
    ten limits silently resolved to `house_default` — a lookup that reads as live
    and is dead.
  * `riskBoard.ts` also once carried the single-name and geo caps TRANSPOSED
    (0.35 / 0.20), judging every position against the wrong ceiling in both
    directions. The comment left behind asks the next person to keep the copies in
    step by hand. This module is the answer to that comment.

So the cap VALUES live here and nowhere else. `book_metrics` re-exports them for
the modules that already import from it, `OptimizerConstraints` is built FROM a
`Mandate` rather than from module constants, and
`frontend/tests/unit/mandate-drift.test.ts` parses this file and fails if the
TypeScript mirror disagrees — the pattern `risk-thresholds.test.ts` already uses
to pin the minimum-sample rule across the language boundary (ADR-0100).

WHAT IS ENFORCED AND WHAT IS ONLY WATCHED
-----------------------------------------
Only the constraints in this file are entered into the solver. The risk board also
renders VaR, CVaR, drawdown, HHI, net exposure and beta limits, and **nothing in
the backend constrains any of them**. They are monitoring thresholds, and the
frontend type keeps them in a separate shape so the two cannot be confused: a
reader looking at a limit is owed the difference between "the sizer cannot breach
this" and "we would like to know if this happens".

PROVENANCE IS PART OF THE VALUE
-------------------------------
`Mandate.sources` records, per field, whether the number came from
`scoring_config` or from the fallback in this module. A mandate panel that shows
35% without saying where 35% came from is a naked number (design goal 1), and the
honest answer differs per field: the caps are seeded, `max_longs` and `lens` are
not yet.

THE VALUES ARE AN OPERATOR DECISION
-----------------------------------
ADR-0037: "The limit *values* (20% single name / 30% sector / 35% geography) are
unchanged. They were never the defect and re-specifying them is an operator
decision, not an implementation one." Changing one is a `scoring_config` write and
takes effect on the next 21:30 UTC run. Nothing in the frontend may write them —
design goal 5.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

# ─── Enforced constraints ────────────────────────────────────────────────────
# These are entered into the solver as constraints (ADR-0107) and clamped by the
# heuristic allocator (ADR-0037). A book cannot breach them.

TOTAL_CAPITAL = 100_000_000.0  # the book is sized against this
MAX_SINGLE_NAME_WEIGHT = 0.20  # no single position > 20% of book
MAX_SECTOR_WEIGHT = 0.30       # no single sector > 30%
MAX_GEO_WEIGHT = 0.35          # no single geography > 35%
MAX_GROSS = 1.0                # long + short <= 100%; ADR-0037 banks the rest as cash

# A CORRELATION COMPLEX IS ONE IDEA, so it may hold at most what one name may
# (ADR-0115), and at most one name's worth of RISK (ADR-0118). Named separately
# from the single-name cap rather than aliased, so the two can diverge later with
# an argument rather than by accident.
MAX_COMPLEX_WEIGHT = 0.20

# A crowded name's single-name cap is HALVED (ADR-0110). Every other name is absent
# from the crowding map and sized bit-identically — absence is neutrality. This may
# only ever tighten a cap, never raise one.
CROWDED_CAP_MULTIPLIER = 0.5

# Book shape. Not currently seeded into scoring_config; `Mandate.sources` says so.
MAX_LONGS = 5
MAX_SHORTS = 5

# Which asset classes the candidate pool is filtered to before the LLM sees it
# (ADR-0015). `multi_asset` is what the nightly run ships; ADR-0015 records that
# Andromeda Capital's own mandate is credit + rates, which is a different lens and
# an operator decision with a real effect on the book.
DEFAULT_LENS = "multi_asset"

LENSES = ("multi_asset", "credit", "rates", "equity", "fx", "commodity")

# ─── scoring_config keys ─────────────────────────────────────────────────────
# The key each field is read from. `riskBoard.ts` already looks up the first three
# of these; until the seeding migration they resolved to nothing.

CONFIG_KEYS: dict[str, str] = {
    "total_capital": "total_capital",
    "max_single_name": "max_single_name_weight",
    "max_sector": "max_sector_weight",
    "max_geo": "max_geo_weight",
    "max_gross": "max_gross",
    "max_complex": "max_complex_weight",
    "crowded_multiplier": "crowded_cap_multiplier",
}

Source = Literal["scoring_config", "code_default"]


@dataclass(frozen=True)
class Mandate:
    """The constraints one book is run under.

    Frozen: a mandate is fixed for the duration of a run. Re-sizing under a
    different mandate constructs a different `Mandate` and calls the same sizer —
    which is what lets the nightly job, the workbench and the MCP `size_book` tool
    share one optimizer without being able to disagree.
    """

    total_capital: float = TOTAL_CAPITAL
    max_single_name: float = MAX_SINGLE_NAME_WEIGHT
    max_sector: float = MAX_SECTOR_WEIGHT
    max_geo: float = MAX_GEO_WEIGHT
    max_gross: float = MAX_GROSS
    max_complex: float = MAX_COMPLEX_WEIGHT
    crowded_multiplier: float = CROWDED_CAP_MULTIPLIER
    max_longs: int = MAX_LONGS
    max_shorts: int = MAX_SHORTS
    lens: str = DEFAULT_LENS
    # Per-field provenance. A field absent from this map came from this module's
    # fallback; `source_of` reports `code_default` for it rather than guessing.
    sources: dict[str, Source] = field(default_factory=dict)

    def source_of(self, name: str) -> Source:
        """Where did this field's value come from? Never guesses."""
        return self.sources.get(name, "code_default")

    @classmethod
    def from_rows(cls, rows: list[dict[str, Any]] | None) -> "Mandate":
        """Build from raw `scoring_config` rows.

        Takes the same shape `ScoringConfig.from_db_rows` takes, so a caller that
        already read the table does not read it twice. A missing or unparseable row
        falls back to this module's value and is recorded as `code_default` — an
        absent limit must not silently become a permissive one.
        """
        vals: dict[str, float] = {}
        for row in rows or []:
            name = row.get("param_name")
            if not name:
                continue
            try:
                vals[name] = float(row["value"])
            except (KeyError, TypeError, ValueError):
                # A malformed row is a missing row. It must not become a limit.
                continue

        kwargs: dict[str, Any] = {}
        sources: dict[str, Source] = {}
        for attr, key in CONFIG_KEYS.items():
            if key in vals:
                kwargs[attr] = vals[key]
                sources[attr] = "scoring_config"

        return cls(**kwargs, sources=sources)

    def to_dict(self) -> dict[str, Any]:
        """Serialisable form, provenance included.

        Consumed by the mandate panel and by the MCP `size_book` tool. Provenance
        travels WITH the value because a limit whose origin a reader cannot name is
        a naked number (design goal 1).
        """
        fields = (
            "total_capital", "max_single_name", "max_sector", "max_geo",
            "max_gross", "max_complex", "crowded_multiplier",
            "max_longs", "max_shorts", "lens",
        )
        return {
            "values": {name: getattr(self, name) for name in fields},
            "sources": {name: self.source_of(name) for name in fields},
            "config_keys": dict(CONFIG_KEYS),
        }


# The mandate the nightly run ships when `scoring_config` is unreachable. Named
# rather than constructed inline so a caller cannot mistake it for a live read.
DEFAULT_MANDATE = Mandate()

File: frontend/lib/mandate.ts
// The mandate: the constraints the book is run under, mirrored from the backend.
//
// The values live in `backend/services/mandate.py`. This file is the frontend's
// reading of them, and `tests/unit/mandate-drift.test.ts` parses that Python and
// fails if the two disagree — the pattern `risk-thresholds.test.ts` already uses to
// pin the minimum-sample rule across the language boundary (ADR-0100).
//
// A hand-maintained copy is what this replaces, and it had drifted twice:
//
//   * the single-name and geo caps shipped TRANSPOSED (0.35 / 0.20), so the risk
//     board judged every position against the wrong ceiling — understating
//     single-name breaches and overstating geographic ones;
//   * `gross_exposure_pct` still read 2.0 (200%) against an optimizer that has
//     enforced `max_gross = 1.0` since ADR-0037. A published risk limit permitting
//     leverage the sizer cannot produce is the defect ADR-0123 names.
//
// ENFORCED vs MONITORED is a real distinction, not a label
// --------------------------------------------------------
// `ENFORCED` limits are entered into the solver as constraints (ADR-0107) and
// clamped by the heuristic allocator (ADR-0037). A published book cannot breach
// one. `MONITORED` thresholds are watched and reported, and **nothing in the
// backend constrains them** — there is no net-exposure or beta constraint anywhere
// in the sizer. Rendering the two in one undifferentiated list told a reader that
// a 30% net band was as binding as the 20% single-name cap. It is not.
//
// They are separate types so the difference cannot be lost by a later edit.

/** A limit the sizer physically cannot breach. */
export interface EnforcedLimit {
  readonly value: number;
  /** The `scoring_config.param_name` this is read from. */
  readonly configKey: string;
  readonly kind: "enforced";
}

/** A threshold that is watched and reported, but constrains nothing. */
export interface MonitoredThreshold {
  readonly value: number;
  readonly configKey: string;
  readonly kind: "monitored";
}

const enforced = (value: number, configKey: string): EnforcedLimit => ({
  value,
  configKey,
  kind: "enforced",
});

const monitored = (value: number, configKey: string): MonitoredThreshold => ({
  value,
  configKey,
  kind: "monitored",
});

/**
 * Constraints the solver enforces. Mirrors `backend/services/mandate.py`.
 *
 * Percentages are decimals (0.20 = 20%). Keep in step with the Python or
 * `mandate-drift.test.ts` fails — which is the point.
 */
export const ENFORCED = {
  /** The book is sized against this. */
  total_capital: enforced(100_000_000, "total_capital"),
  /** No single position > 20% of book. */
  single_name_pct: enforced(0.2, "max_single_name_weight"),
  /** No single sector > 30%. */
  sector_pct: enforced(0.3, "max_sector_weight"),
  /** No single geography > 35%. */
  geo_pct: enforced(0.35, "max_geo_weight"),
  /**
   * Long + short <= 100%. NOT 200%: ADR-0037 holds whatever the position limits
   * refuse as cash rather than renormalising the book back to full notional, so
   * `gross <= 1` is a ceiling the sizer reaches from below and never exceeds.
   */
  gross_exposure_pct: enforced(1.0, "max_gross"),
  /** A correlation complex is one idea, so it holds at most what one name may (ADR-0115). */
  complex_pct: enforced(0.2, "max_complex_weight"),
  /** A crowded name's single-name cap is halved (ADR-0110). Only ever tightens. */
  crowded_multiplier: enforced(0.5, "crowded_cap_multiplier"),
  /**
   * Day-over-day, against yesterday's PUBLISHED book — distinct from the intra-run
   * figure (this same solve vs the conviction book), which nothing constrains.
   * ADR-0107 carried this field since the optimizer shipped and nothing ever set
   * it; ADR-0173 is the first value. A first cut, not a fitted optimum — chosen
   * above the typical 30–58% band so an ordinary day is unaffected, and well below
   * the 72–200% band the two worst observed days sat in.
   */
  turnover_pct: enforced(0.6, "max_turnover"),
} as const;

/**
 * Thresholds the risk board reports against. **None of these constrains the sizer.**
 *
 * A book can and does breach these without any code preventing it — they exist so a
 * reader learns that it happened, which is a different promise from the caps above
 * and must read as one.
 */
export const MONITORED = {
  var_95_pct: monitored(0.06, "limit_var_95_pct"),
  cvar_95_pct: monitored(0.09, "limit_cvar_95_pct"),
  max_drawdown_pct: monitored(0.15, "limit_max_drawdown_pct"),
  /** ±30% net long/short of capital. Watched only — there is no net constraint in the solver. */
  net_exposure_pct: monitored(0.3, "limit_net_exposure_pct"),
  /** |beta| to SPX. Watched only — nothing in the backend targets a beta. */
  beta_abs: monitored(0.5, "limit_beta_abs"),
  /**
   * Herfindahl ceiling on the 0–10 000 (DOJ) scale — backend
   * `concentration_hhi = Σwᵢ²·10 000`, so the limit must be on the same scale.
   * 2 000 ≈ five equal-weight names; an earlier 0–1-scale 0.2 made a 2 500 book
   * read as 1 250 000 % utilisation.
   */
  hhi: monitored(2000, "limit_hhi"),
} as const;

/** Book shape. Not seeded into `scoring_config` — provenance reports `code_default`. */
export const BOOK_SHAPE = {
  max_longs: 5,
  max_shorts: 5,
} as const;

/** The lens the nightly run ships. ADR-0015 records that the firm's own mandate is credit + rates. */
export const DEFAULT_LENS = "multi_asset";

export const LENSES = [
  "multi_asset",
  "credit",
  "rates",
  "equity",
  "fx",
  "commodity",
] as const;

export type EnforcedKey = keyof typeof ENFORCED;
export type MonitoredKey = keyof typeof MONITORED;

/** Where a rendered limit's value actually came from. */
export type LimitSource = "scoring_config" | "code_default";

/**
 * Every limit as one flat list, enforcement status attached.
 *
 * For a mandate panel that must show all of them together while still saying which
 * ones bind. Deliberately not a merge into one shape — `kind` travels with each row.
 */
export function allLimits(): Array<
  (EnforcedLimit | MonitoredThreshold) & { key: string }
> {
  return [
    ...Object.entries(ENFORCED).map(([key, l]) => ({ ...l, key })),
    ...Object.entries(MONITORED).map(([key, l]) => ({ ...l, key })),
  ];
}
